### PR TITLE
[SPARK-42774][SQL]Expose VectorTypes API for DataSourceV2 Batch Scans

### PR DESCRIPTION
**Pull request by xiaochen-zhou**
What changes were proposed in this pull request?
Add an override to BatchScanExecBase which delegates to a new default method on PartitionReaderFactory to expose vectoryTypes.

Why are the changes needed?
SparkPlan's vectorType's attribute can be used to specialize codegen however BatchScanExecBase does not override this so we DSv2 sources do not get any benefit of concrete class dispatch.

Does this PR introduce any user-facing change?
NO

How was this patch tested?
Pass GitHub Actions

### CODE REVIEW

File: sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaColumnarDataSourceV2.java [hunk 1]
```diff
@@ -69,6 +72,14 @@ public boolean supportColumnarReads(InputPartition partition) {
       return true;
     }
 
+    @Override
+    public Optional<Iterable<String>> getVectorTypes() {
+      List<String> vectorTypes = new ArrayList<>();
+      vectorTypes.add(OnHeapColumnVector.class.getName());
+      vectorTypes.add(OnHeapColumnVector.class.getName());
```
Comment (jaceklaskowski) [inline sql/core/src/test/java/test/org/apache/spark/sql/connector/JavaColumnarDataSourceV2.java:79]: nit: [Collections.nCopies(int, Object)](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#nCopies-int-T-)

File: sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala [hunk 2]
```diff
@@ -905,6 +908,13 @@ object ColumnarReaderFactory extends PartitionReaderFactory {
 
   override def supportColumnarReads(partition: InputPartition): Boolean = true
 
+  override def getVectorTypes: Optional[java.lang.Iterable[String]] = {
+    val data: Iterable[String] = Iterable.fill(2)(
+        classOf[OnHeapColumnVector].getName
+    )
```
Comment (jaceklaskowski) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:914]: Oh, that's also very nice (re my previous comment to use `Collections.nCopies`).

Can you rename `data`?

Comment (clownxc) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/DataSourceV2Suite.scala:914]: > Oh, that's also very nice (re my previous comment to use `Collections.nCopies`).
> 
> Can you rename `data`?

Thank you, I will try to modify the code as you say

**Comment by clownxc** [comment]
I applied the suggested code changes.Can you please do another review @jaceklaskowski

**Comment by clownxc** [comment]
Could you please take a look when you have time? thanks! @cloud-fan

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java [hunk 3]
```diff
@@ -37,15 +37,15 @@
 public interface PartitionReaderFactory extends Serializable {
 
   /**
-   * Returns a row-based partition reader to read data from the given {@link InputPartition}.
+   * @return a row-based partition reader to read data from the given {@link InputPartition}.
    * <p>
    * Implementations probably need to cast the input partition to the concrete
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:42]: Can we add `@note` here so `@return` is just one-liner?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java [hunk 4]
```diff
@@ -37,15 +37,15 @@
...
    * {@link InputPartition} class defined for the data source.
    */
   PartitionReader<InternalRow> createReader(InputPartition partition);
 
   /**
-   * Returns a columnar partition reader to read data from the given {@link InputPartition}.
+   * @return a columnar partition reader to read data from the given {@link InputPartition}.
    * <p>
    * Implementations probably need to cast the input partition to the concrete
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:50]: same here

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java [hunk 5]
```diff
@@ -55,7 +55,7 @@ default PartitionReader<ColumnarBatch> createColumnarReader(InputPartition parti
   }
 
   /**
-   * Returns true if the given {@link InputPartition} should be read by Spark in a columnar way.
+   * @return true if the given {@link InputPartition} should be read by Spark in a columnar way.
    * This means, implementations must also implement {@link #createColumnarReader(InputPartition)}
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:59]: and here

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala [hunk 6]
```diff
@@ -65,7 +63,11 @@ trait DataSourceV2ScanExecBase extends LeafExecNode {
   }
 
   override def vectorTypes: Option[Seq[String]] = {
-    Option(readerFactory.getVectorTypes.get.asScala.toSeq)
+    val vectorTypes = readerFactory.getVectorTypes
+    if (vectorTypes.isPresent) {
+      vectorTypes.get()
```
Comment (jaceklaskowski) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala:68]: What does this line do? Looks like a missing `return`?

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java [hunk 7]
```diff
@@ -68,7 +68,7 @@ default boolean supportColumnarReads(InputPartition partition) {
   }
 
   /**
-   * Returns exact java types of the columns that are output in columnar processing mode.
+   * @return exact java types of the columns that are output in columnar processing mode.
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:71]: "Fully-qualified Java class names of"?

**Review by jaceklaskowski — COMMENTED** [review]
nit-picking 🤷

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java (hunk 5 above)
Comment (clownxc) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:59]: > and here

Thank you very much for review and sorry for the late response. I will try to modify the code according to your suggestions

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java (hunk 7 above)
Comment (clownxc) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:72]: > "Fully-qualified Java class names of"?

like this `org.apache.spark.sql.execution.vectorized.OnHeapColumnVector`

**Comment by clownxc** [comment]
I applied the suggested code changes. can you review the code one more time in your spare time, thanks a lot @jaceklaskowski

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java (hunk 7 above)
Comment (jaceklaskowski) [reply sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:72]: `@return The fully-qualified Java class names of...`

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java [hunk 8]
```diff
@@ -36,33 +37,42 @@
 public interface PartitionReaderFactory extends Serializable {
 
   /**
-   * Returns a row-based partition reader to read data from the given {@link InputPartition}.
+   * @return a row-based partition reader to read data from the given {@link InputPartition}.
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:40]: nit: A (uppercase)

File: sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java [hunk 9]
```diff
@@ -36,33 +37,42 @@
...
-   * Implementations probably need to cast the input partition to the concrete
-   * {@link InputPartition} class defined for the data source.
+   * @note Implementations probably need to cast the input partition to the concrete
+   *       {@link InputPartition} class defined for the data source.
    */
   PartitionReader<InternalRow> createReader(InputPartition partition);
 
   /**
-   * Returns a columnar partition reader to read data from the given {@link InputPartition}.
+   * @return a columnar partition reader to read data from the given {@link InputPartition}.
```
Comment (jaceklaskowski) [inline sql/catalyst/src/main/java/org/apache/spark/sql/connector/read/PartitionReaderFactory.java:48]: Same here

**Pull request closed without merging**